Add a page on CSRF (#38151)
mdn/content: files/en-us/web/security/attacks/csrf/index.md

---
title: Cross-site request forgery (CSRF)
slug: Web/Security/Attacks/CSRF
page-type: guide
---

> In a cross-site request forgery (CSRF) attack, an attacker tricks the user or the browser into making an HTTP request to the target site. The request includes the user's credentials and causes the server to carry out some harmful action, thinking that the user intended it.

Reviewer: Can we make it clear this is actually a cross site request, not XSS where its injecting malicious code in the same-site context? This is same comment as https://github.com/mdn/content/pull/38151/files#r1957450793 -

Author: https://github.com/mdn/content/pull/38151/commits/2b4c1346c94fbfd44359282ee47eb7a646e49df2

## Overview

> A website typically performs special actions on a user's behalf — buying a product or making an appointment, for example — by receiving an HTTP request from the user's browser, often with URL parameters detailing the action to perform. To ensure that the request really comes from the user in question, the server expects the request to include {{glossary("Credential", "credentials")}} for the user: for example, a cookie containing the user's session ID.

Reviewer: With CRSF tokens at least we're concerned about POST requests. More generally we're not supposed to ever send _URL_ parameters in state changing requests. See also comment about the "Post in the image"

```suggestion
A website typically performs special actions on a user's behalf — buying a product or making an appointment, for example — by receiving an HTTP request from the user's browser, often with parameters detailing the action to perform. To ensure that the request really comes from the user in question, the server expects the request to include {{glossary("Credential", "credentials")}} for the user: for example, a cookie containing the user's session ID.
```

> In the example below, the user has previously signed into their bank, and the browser has stored a session cookie for the user. The page contains a {{htmlelement("form")}} element, which enables the user to transfer funds to another person. When the user submits the form, the browser sends a {{httpmethod("POST")}} request to the server, including the form data. If the user is signed in, the request includes the user's cookie. The server validates the cookie and performs the special action — in this case, transferring money:

Reviewer: There is nothing wrong with this except it is the "good" or "intended flow" for which we will then compare the bad behaviour. I'd spell it out.

```suggestion
The example below shows the expected interactions when a user signs into their bank using a session cookies stored in the browser. The page contains a {{htmlelement("form")}} element, which enables the user to transfer funds to another person. When the user submits the form, the browser sends a {{httpmethod("POST")}} request to the server, including the form data. If the user is signed in, the request includes the user's cookie. The server validates the cookie and performs the special action — in this case, transferring money:
```

Author: Oh I just accepted this but it is wrong. It doesn't show "the expected interactions when a user signs into their bank using a session cookie stored in the browser". It shows the expected interaction when the user sends a state-changing request _having previously signed in_".

I'll push an update.

> ![Diagram showing a user submitting a browser form, the browser then making a POST request to the server, and the server validating the request.](form-post.svg)

Reviewer: Is this just shorthand? The URL is an HTTP get

![image](https://github.com/user-attachments/assets/052341cd-88fe-41dc-ac55-b368191546c2)

Author: oops, yes you are right, this ought to be fixed.

Author: -> https://github.com/mdn/content/pull/38151/commits/2b4c1346c94fbfd44359282ee47eb7a646e49df2

In this guide, we'll call a request like this, that carries out some special action, a _state-changing request_.

In a CSRF attack, the attacker creates a website containing a form. The form's [`action` attribute](/en-US/docs/Web/HTML/Element/form#action) is set to the bank's website, and the form contains hidden input fields mimicking the bank's fields:

```html
<form action="https://my-bank.example.org/transfer" method="POST">
  <input type="hidden" name="recipient" value="attacker" />
  <input type="hidden" name="amount" value="1000" />
</form>
```

The page also contains JavaScript that submits the form on page load:

```js
const form = document.querySelector("form");
form.submit();
```

> When the user visits the page, the browser submits the form to the bank's website. Because the user is signed into their bank, the request includes the user's real cookie, so the bank's server successfully validates the request, and transfers the funds:

Reviewer: A little bit of a mouthful. I've also used "If the request" - mostly because this is not the default (Lax is, usually), and the cookie isn't sent.

```suggestion
When the user visits the page, the browser submits the form to the bank's website. If the request includes the user's sign-in cookie, the bank's server will successfully validate the request, and transfers the funds:
```

Author: I think the causality is important here. **Because** the user is signed into their bank, the request (might) include the user's session cookie for the bank.

Author: I did though put "include"->"may include" in https://github.com/mdn/content/pull/38151/commits/2b4c1346c94fbfd44359282ee47eb7a646e49df2.

Reviewer: For whatever reason, I'm no longer finding this problematic.

![Diagram showing a CSRF attack in which a decoy page submits a POST request to the website for the user's bank.](csrf-form-post.svg)

> There are other ways the attacker could issue a cross-site request forgery. For example, if the website uses a {{httpmethod("GET")}} request to carry out the action, then the attacker can avoid having to use a form at all, and can execute the attack by sending the user a link to a page that contains markup like this:

Reviewer: It might not be worth saying here but it should be said in this doc (I'm working down, so apologies if you already covered it), but as above, GET should never be used for state changing requests for exactly this reason - too easy to hack, and you can't use CRSF tokens in this case.

Author: Yes, I agree about this but I'm not sure where to put it exactly. I mentioned it in the bit about SameSite since it's very relevant to Lax.

And yes, at least for Django you need to avoid `GET` if you want to use CSRF tokens, and we could mention it there too, but that's really just a functional point about implementing that defense, and it's already covered in the Django docs for that (https://docs.djangoproject.com/en/5.1/ref/csrf/), so it doesn't feel like an "extra defense".

Mentioning it on its own section under "defenses" seems wrong too, since it's not a defense on its own.

I'll think some more about it.

```html
<img
  src="https://my-bank.example.org/transfer?recipient=attacker&amount=1000" />
```

When the user loads the page, the browser tries to fetch the image resource, which is really the transaction request.

In general, a CSRF attack is possible if your website:

- Uses HTTP requests to change some state on the server.
- Uses cookies to validate that the request came from an authenticated user.
- Uses parameters in the request that an attacker can predict.

## Defenses against CSRF

In this section we'll outline two alternative defenses against CSRF and a third practice which can be used to provide defense in depth for either of the other two.

> - The first primary defense is to use _CSRF tokens_ embedded in the page. This method is especially appropriate if you're issuing state-changing requests from form elements, as in our example above.

Reviewer: Perhaps 

```suggestion
- The first primary defense is to use _CSRF tokens_ embedded in the page. This method is the defacto-standard approach when issuing state-changing requests from form elements, as in our example above.
```

Author: -> https://github.com/mdn/content/pull/38151/commits/2b4c1346c94fbfd44359282ee47eb7a646e49df2

> - The alternative defense is to ensure that state-changing requests are not _simple requests_, which enables them to rely on the browser's built-in cross-origin request blocking. This method is appropriate if you're issuing state-changing requests from JavaScript APIs like {{domxref("Window.fetch()", "fetch()")}}.

Reviewer: Perhaps more direct?

```suggestion
- The alternative defense is to ensure that state-changing requests are not CORS _simple requests_, ensuring that cross-origin requests are blocked by default. This method is appropriate if you're issuing state-changing requests from JavaScript APIs like {{domxref("Window.fetch()", "fetch()")}}.
```

Author: Perhaps "ensuring that" -> "which ensures that", to make it clear that the blocking is a consequence of the non-simpleness, and is not an extra thing people have to do?

Reviewer: Perhaps, though since I dislike the double-ensure use, perhaps

```suggestion
- The alternative defense is to ensure that state-changing requests are not CORS _simple requests_, so that cross-origin requests are blocked by default. This method is appropriate if you're issuing state-changing requests from JavaScript APIs like {{domxref("Window.fetch()", "fetch()")}}.
```

Author: Mostly accepted in https://github.com/mdn/content/pull/38151/commits/2b4c1346c94fbfd44359282ee47eb7a646e49df2, although I didn't add "CORS" because I want to put CORS in the back seat here. A lot of people think CORS is a security feature, and I want to make it as clear as possible that what we're talking about here is the default behavior, with no CORS.

Finally, we'll discuss the `SameSite` cookie attribute, which can be used to provide defense in depth alongside either of the previous methods.

### CSRF tokens

> In this defense, when the server serves a page, it embeds an unpredictable value in the page, called the CSRF token. Then when the browser sends the state-changing request to the server, it includes the CSRF token in the HTTP request. The browser checks the token value and carries out the request only if it matches. Because an attacker can't guess the token value, they can't issue a successful forgery.

Reviewer: Worth commenting on the replay-preventing aspects of this? Something like

```suggestion
In this defense, when the server serves a page, it embeds an unpredictable value in the page, called the CSRF token. Then when the browser sends the state-changing request to the server, it includes the CSRF token in the HTTP request. The browser checks the token value and carries out the request only if it matches. Because an attacker can't guess the token value, they can't issue a successful forgery. Even if the attacker does somehow discover a token at a later point, the request can't be replayed if the token changes every time! 
```

Author: -> https://github.com/mdn/content/pull/38151/commits/2b4c1346c94fbfd44359282ee47eb7a646e49df2

> For form submissions, the CSRF token is usually implemented as a hidden form field. For a JavaScript API like `fetch()`, the token might be placed in a cookie or embedded in the page, and the JavaScript extracts the value and sends it as an extra header.

Author: AFAIK there is no standard here but happy to be corrected.

> For form submissions, the CSRF token is usually implemented as a hidden form field. For a JavaScript API like `fetch()`, the token might be placed in a cookie or embedded in the page, and the JavaScript extracts the value and sends it as an extra header.

Reviewer: I don't know either. Entirely up to you, but you could say as a `server-specific header`

Modern web frameworks usually have built-in support for CSRF tokens: for example, [Django](https://www.djangoproject.com/) enables you to protect forms using the [`csrf_token`](https://docs.djangoproject.com/en/5.1/ref/csrf/) tag. This generates an additional hidden form field containing the token, which the framework then checks on the server.

To take advantage of this protection you must understand all the places in your website where you are using state-changing HTTP requests, and ensure you're using the defense provided by your chosen framework.

### Avoiding simple requests

Web browsers distinguish two sorts of HTTP requests: [_simple_ requests](/en-US/docs/Web/HTTP/CORS#simple_requests) and other requests.

> By default, only simple requests can be sent cross-origin. The reason for allowing simple requests cross-origin is that these are the same sorts of requests that could already be made cross-origin using a `<form>` element, as in the example above. So it is assumed that websites must already implement CSRF protection against simple requests.
> 
> All other requests are by default not allowed cross-origin, so a CSRF attack would not succeed if the request is not simple.
> 
> So one CSRF defense is to ensure that state-changing requests are never simple requests. This of course means that a website can't use forms to issue them, so this strategy is usually applicable for a website that uses JavaScript APIs like {{domxref("Window.fetch()", "fetch()")}} to issue state-changing requests.

Author: So perhaps we could say something like:

```suggestion
Simple requests are the sorts of requests that a `<form>` element can make. Since forms have been able to make cross-origin requests since the early days of the web, it's important for compatibility that they should still be able to make cross-origin requests, and this means that forms must defend against CSRF using some mechanism such as a CSRF token.

However, other parts of the web platform, in particular JavaScript APIs like {{domxref("Window.fetch()", "fetch()")}}, can make different sorts of requests (for example, requests that set custom headers), and these requests are by default not allowed cross-origin, so a CSRF attack would not succeed.

So a website that uses `fetch()` or `XMLHttpRequest` can defend against CSRF by ensuring that the state-changing requests that it issues are never simple requests.
```

?

> So one CSRF defense is to ensure that state-changing requests are never simple requests. This of course means that a website can't use forms to issue them, so this strategy is usually applicable for a website that uses JavaScript APIs like {{domxref("Window.fetch()", "fetch()")}} to issue state-changing requests.

Author: > So it is assumed that websites must already implement CSRF protection against simple requests." is quite true either - I suspect it would be a non-simple request if there wasn't prior art meaning that too much of the web would break.

Well yes, of course the reason cross-origin form requests are allowed is historical: by the time it was obvious that there was a problem here, there were too many legacy sites doing it for it to be fixed.

But it's _because of that_ that we must assume people using forms are also implementing their own CSRF - we just have to assume that, because we can't fix it in the browser. This is covered quite explicitly in the [CORS docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#simple_requests) which I link in the previous para:

> The motivation is that the [`<form>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form) element from HTML 4.0 (which predates cross-site [fetch()](https://developer.mozilla.org/en-US/docs/Web/API/Window/fetch) and [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest)) can submit simple requests to any origin, so anyone writing a server must already be protecting against [cross-site request forgery](https://developer.mozilla.org/en-US/docs/Glossary/CSRF) (CSRF). Under this assumption, the server doesn't have to opt-in (by responding to a preflight request) to receive any request that looks like a form submission, since the threat of CSRF is no worse than that of form submission.

So I'm not that keen on "unfortunately", since it implies that this is an unhappy accident, that simple requests just happen to include form submissions. But actually AIUI, by definition really, simple requests are the ones that forms can make (we might even call them "form-compatible requests"). If forms had made different requests, then the definition of a simple request would have been correspondingly different.

> So one CSRF defense is to ensure that state-changing requests are never simple requests. This of course means that a website can't use forms to issue them, so this strategy is usually applicable for a website that uses JavaScript APIs like {{domxref("Window.fetch()", "fetch()")}} to issue state-changing requests.

Author: So I made a different edit in https://github.com/mdn/content/pull/38151/commits/2b4c1346c94fbfd44359282ee47eb7a646e49df2.

> So one CSRF defense is to ensure that state-changing requests are never simple requests. This of course means that a website can't use forms to issue them, so this strategy is usually applicable for a website that uses JavaScript APIs like {{domxref("Window.fetch()", "fetch()")}} to issue state-changing requests.

Reviewer: I like the way you dealt with this.

#### Non-simple requests and CORS

We've said that non-simple requests are _by default_ not sent cross-origin. The catch is that the [Cross-Origin Resource Sharing (CORS)](/en-US/docs/Web/HTTP/CORS) protocol allows a website to relax this restriction.

When a site issues a non-simple request cross-origin, the browser sends a {{glossary("Preflight request", "preflight request")}}. This asks the server if it is prepared to accept the request.

- If the server responds with an {{httpheader("Access-Control-Allow-Origin")}} response header which matches the sender's origin, then the browser will send the real request.
- If the server also sends an {{httpheader("Access-Control-Allow-Credentials")}} response header, then the real request is allowed to include the site's credentials.

This is obviously useful in cases where you want to accept requests from some other origins. However, it means that if your server sends an `Access-Control-Allow-Origin` response header including the sender's origin, and an `Access-Control-Allow-Credentials` response header, then the server is vulnerable to a CSRF attack from that origin.

### Defense in depth: SameSite cookies

> 

Reviewer: I like this. In particular the potential issues with Lax have never been explained to me before.

Author: Actually I feel this is vague and wish I had a concrete example of how an attacker could circumvent `Lax`. I just [copied this bit from the spec](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-02#section-5.3.7.1) but don't really understand it. Don't tell anyone that though.

Reviewer: It is a big vague. I think only top level navigations with GET include cookies, so you'd be mostly fine if you don't do something dumb like make state-changing  requests using URL params.

Here is some reading https://portswigger.net/web-security/csrf/bypassing-samesite-restrictions#bypassing-samesite-lax-restrictions-using-get-requests

There are a few cases worth highlighting there - samesite isn't sameorigin.

Reviewer: I.e. I don't see how the popup case is a bypass or in some way an avenue for a same-site attack unless it gets to you maybe to open a subdomain - i.e. same-origin.

Author: > It is a big vague. I think only top level navigations with GET include cookies, so you'd be mostly fine if you don't do something dumb like make state-changing requests using URL params.

Seems like it's not just GET, but is only [safe](https://developer.mozilla.org/en-US/docs/Glossary/Safe/HTTP) methods: https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-02#section-5.5. And yes, the other bit of the spec does say [SameSite gives you good protection if you enforce the use of unsafe methods like POST](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-02#section-5.3.7.1). So perhaps we can use this as a place to say don't issue state-changing requests using unsafe methods, and here's a reason why.

And if you do only use unsafe methods, then SameSite is a reasonable defense, except for the samesite != sameorigin thing.

> 
> Here is some reading https://portswigger.net/web-security/csrf/bypassing-samesite-restrictions#bypassing-samesite-lax-restrictions-using-get-requests
> 

Ah, that's a helpful link, yes.

Reviewer: Yes, - safe methods are this list: 

> "Safe" HTTP methods include "GET", "HEAD", "OPTIONS", and "TRACE", as
   defined in [Section 4.2.1 of [RFC7231]](https://datatracker.ietf.org/doc/html/rfc7231#section-4.2.1).
   
So the methods you might use for changing something is GET. There is another note in the spect that safe method implementations should be kept idempotent.

Author: I pushed https://github.com/mdn/content/pull/38151/commits/5a6d295e125aa3f4c60d9f4ccf6f698627a567ae to go into some more detail on SameSite issues.

The `Lax` value relaxes this restriction: cookies are included in cross-site requests for top-level navigations only. This means that, for example, attempts to make a request using a form submission would not include `Lax` cookies. The problem with this value is that attackers can still pop up new windows or trigger top-level
navigations in order to create cross-site requests that satisfy the `Lax` constraint.

Another problem with the `SameSite` attribute is that it protects you from requests from a different {{glossary("Site", "site")}}, not a different {{glossary("Origin", "origin")}}. This is a looser protection, because (for example) `https://foo.example.org` and `https://bar.example.org` are considered the same site, although they are different origins. Effectively, if you rely on same-site protection, you have to trust all your site's subdomains.

Even so, it is worth setting the `SameSite` attribute for sensitive cookies to `Strict` if you can, or `Lax` if you have to.

> 

Reviewer: Perhaps a "[Defense summary checklist](https://developer.mozilla.org/en-US/docs/Web/Security/Attacks/XSS#defense_summary_checklist)" like that one.

Author: -> https://github.com/mdn/content/pull/38151/commits/60a0284ceb7a739ad063b118f842675cad099679

For form submissions, the CSRF token is usually included in a hidden form field, so that on form submission it is automatically sent back to the server for checking.

For a JavaScript API like `fetch()`, the token might be placed in a cookie or embedded in the page, and the JavaScript extracts the value and sends it as an extra header.

Modern web frameworks usually have built-in support for CSRF tokens: for example, [Django](https://www.djangoproject.com/) enables you to protect forms using the [`csrf_token`](https://docs.djangoproject.com/en/5.1/ref/csrf/) tag. This generates an additional hidden form field containing the token, which the framework then checks on the server.

To take advantage of this protection you must understand all the places in your website where you are using state-changing HTTP requests, and ensure you're using the defense provided by your chosen framework.

> ### Avoiding simple requests

Author: OK, I have had a go at this in https://github.com/mdn/content/pull/38151/commits/d425da4b6fe08d753a357485dd67140699b85149. It's unfortunate that we don't have a guide to Fetch metadata on MDN, but perhaps when we have written one, we can revisit this.

So a website that uses `fetch()` or `XMLHttpRequest` can defend against CSRF by ensuring that the state-changing requests that it issues are never simple requests.

For example, setting the request's {{httpheader("Content-Type")}} to `"application/json"` will prevent it from being treated like a simple request:

```js
fetch("https://my-bank.example.org/transfer", {
  method: "POST",
  headers: {
    "Content-Type": "application/json",
  },
  body: JSON.stringify({ recipient: "joe", amount: "100" }),
});
```

Similarly, setting a custom header on the request will prevent it being treated like a simple request:

```js
fetch("https://my-bank.example.org/transfer", {
  method: "POST",
  headers: {
    "X-MY-BANK-ANTI-CSRF": 1,
  },
  body: JSON.stringify({ recipient: "joe", amount: "100" }),
});
```

The header name can be anything, as long as it doesn't conflict with standard headers.

The server can then check for the existence of the header: if it exists, then the server knows that the request was not treated as a simple request.

> #### Non-simple requests and CORS

Author: Yes, I could definitely trim this section. I think it is worth saying "if you rely on this defense, be extra-careful with CORS" but we don't need to explain how it works here.

- If the server responds with an {{httpheader("Access-Control-Allow-Origin")}} response header which matches the sender's origin, then the browser will send the real request.
- If the server also sends an {{httpheader("Access-Control-Allow-Credentials")}} response header, then the real request is allowed to include the site's credentials.

This is obviously useful in cases where you want to accept requests from some other origins. However, it means that if your server sends an `Access-Control-Allow-Origin` response header including the sender's origin, and an `Access-Control-Allow-Credentials` response header, then the server is vulnerable to a CSRF attack from that origin.

### Defense in depth: SameSite cookies

The [`SameSite`](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-02#section-5.3.7) cookie attribute provides some protection against CSRF attacks. It's not a complete defense, and is best considered as an addition to one of the other defenses, providing some degree of defense in depth.

This attribute controls when a browser is allowed to include the cookie in a cross-site request. It has three possible values: `None`, `Lax` and `Strict`.

> The `Strict` value offers the most protection: if this attribute is set, the browser will not include the cookie in any cross-site request. However, this creates a usability issue: if the user is logged into your site, and follows a link to your site from a different site, then your cookies will not be included, and the user will not be recognized when they reach your site.

Author: -> https://github.com/mdn/content/pull/38151/commits/7482ccee91642f167b720f469e8bcac3b538bf14

> The `Strict` value offers the most protection: if this attribute is set, the browser will not include the cookie in any cross-site request. However, this creates a usability issue: if the user is logged into your site, and follows a link to your site from a different site, then your cookies will not be included, and the user will not be recognized when they reach your site.

Reviewer 2: https://github.com/mdn/content/blob/7482ccee91642f167b720f469e8bcac3b538bf14/files/en-us/web/security/attacks/csrf/index.md?plain=1#L127 links to the RFC-draft-that-has-been-undergoing-maintenance-for-many-years-now. Is there nothing on MDN it could link to?

> The `Strict` value offers the most protection: if this attribute is set, the browser will not include the cookie in any cross-site request. However, this creates a usability issue: if the user is logged into your site, and follows a link to your site from a different site, then your cookies will not be included, and the user will not be recognized when they reach your site.

Author: -> https://github.com/mdn/content/pull/38151/commits/4def96ce3533d06a08659214ada517226d5e87cd - use https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Set-Cookie#samesitesamesite-value instead.

The `Lax` value relaxes this restriction: cookies are included in cross-site requests if both the following conditions apply:

- The request was a navigation of the top-level browsing context.
- The request used a {{glossary("Safe/HTTP", "safe")}} method: notably, {{httpmethod("GET")}} is safe but {{httpmethod("POST")}} is not.

However, `Lax` offers significantly weaker protection than `Strict`:

> - An attacker can trigger a top-level navigation that sends the request, for example by setting {{domxref("Document.location", "document.location")}}.

Reviewer: ```suggestion
- An attacker can trigger a top-level navigation, for example by setting {{domxref("Document.location", "document.location")}}, and the resulting `GET` request will still include the session cookie.
```

Reviewer: ^^^^ I wasn't sure how this helped attack. It helps because the GET still sends the cookie, but not if you avoid state changing gets.

> - An attacker can trigger a top-level navigation, for example by setting {{domxref("Document.location", "document.location")}}, and the resulting `GET` request will still include the session cookie.

Author: -> https://github.com/mdn/content/pull/38151/commits/c624ffebbd05c07a90c25ab04b0524db4646363c.

> - Even if the server does check that the request was not sent using `GET`, some web frameworks support "method override": this enables an attacker to send a request using `GET` but have it appear to the server as if it used `POST`.

Author: There is a link to more details on all these issues at the end of the section (https://portswigger.net/web-security/csrf/bypassing-samesite-restrictions).

> So although `Lax` is a reasonable defense in depth against CSRF, as long as you avoid making state-changing requests that use `GET`, it should not be the only defense. It should be deployed alongside one of the previous defenses, as a defense in depth.

Author: ```suggestion
So although `Lax` is a reasonable defense in depth against CSRF, as long as you avoid making state-changing requests that use `GET`, it should not be the only defense. It should be deployed alongside one of the previous defenses.
```

There are other ways the attacker could issue a cross-site request forgery. For example, if the website uses a {{httpmethod("GET")}} request to carry out the action, then the attacker can avoid having to use a form at all, and can execute the attack by sending the user a link to a page that contains markup like this:

```html
<img
  src="https://my-bank.example.org/transfer?recipient=attacker&amount=1000" />
```

When the user loads the page, the browser tries to fetch the image resource, which is really the transaction request.

In general, a CSRF attack is possible if your website:

- Uses HTTP requests to change some state on the server.
- Uses only cookies to validate that the request came from an authenticated user.
- Uses only parameters in the request that an attacker can predict.

## Defenses against CSRF

In this section we'll outline three alternative defenses against CSRF and a fourth practice which can be used to provide defense in depth for either of the others.

- The first primary defense is to [use _CSRF tokens_](#csrf_tokens) embedded in the page. This is the most common method if you're issuing state-changing requests from form elements, as in our example above.

> - The second is to [use _Fetch metadata_](fetch_metadata) HTTP headers to check whether or not the start-changing request is being issued cross-site.

Author: ```suggestion
- The second is to [use _Fetch metadata_](fetch_metadata) HTTP headers to check whether or not the state-changing request is being issued cross-site.
```

For form submissions, the CSRF token is usually included in a hidden form field, so that on form submission it is automatically sent back to the server for checking.

For a JavaScript API like `fetch()`, the token might be placed in a cookie or embedded in the page, and the JavaScript extracts the value and sends it as an extra header.

Modern web frameworks usually have built-in support for CSRF tokens: for example, [Django](https://www.djangoproject.com/) enables you to protect forms using the [`csrf_token`](https://docs.djangoproject.com/en/5.1/ref/csrf/) tag. This generates an additional hidden form field containing the token, which the framework then checks on the server.

To take advantage of this protection you must understand all the places in your website where you are using state-changing HTTP requests, and ensure you're using the defense provided by your chosen framework.

### Fetch metadata

Fetch metadata is a collection of HTTP request headers, added by the browser, that provide extra information about the context of an HTTP request. The server can use these headers to decide whether to allow a request or not.

> Most relevant for CSRF is the {{httpheader("Sec-Fetch-Site")}} header, which tells the server whether this request is same-origin, same-site, cross-site, or initiated directly by the user. For example, this [Express](/en-US/docs/Learn_web_development/Extensions/Server-side/Express_Nodejs) code allows only same-site and same-origin requests:

Reviewer: It might be good to "spell it out" as to why this is useful in the context of CSRF. Just a thought

```suggestion
Most relevant for CSRF is the {{httpheader("Sec-Fetch-Site")}} header, which tells the server whether this request is same-origin, same-site, cross-site, or initiated directly by the user. The server can use this information to allow cross-origin requests, or block them as potential CSRF attacks.

For example, this [Express](/en-US/docs/Learn_web_development/Extensions/Server-side/Express_Nodejs) code allows only same-site and same-origin requests:
```

```js
app.post("/transfer", (req, res) => {
  const secFetchSite = req.headers["sec-fetch-site"];
  if (secFetchSite === "same-origin" || secFetchSite === "same-site") {
    console.log("allowed");
    // Update state
  } else {
    console.log("denied");
    // Don't update state
  }
});
```

See {{glossary("Fetch metadata request header")}} for the complete list of Fetch metadata headers, and [Protect your resources from web attacks with Fetch Metadata](https://web.dev/articles/fetch-metadata) for a guide to using this feature.

### Avoiding simple requests

Web browsers distinguish two sorts of HTTP requests: [_simple_ requests](/en-US/docs/Web/HTTP/CORS#simple_requests) and other requests.

> Simple requests are the sorts of requests that a `<form>` element can make. Since forms have been able to make cross-origin requests since the early days of the web, it's important for compatibility that they should still be able to make cross-origin requests, and this means that forms must defend against CSRF using a mechanism such as a CSRF token.

Reviewer: Bit of a nit, but at the point we talk about form we haven't explained that simple requests are a problem because they allow cross origin requests by default. If it were me I'd make the clarifications about form elements here a note, but I know you don't like those ... how about.

```suggestion
Simple requests, which are the sort of request that result from a `<form>` element submission, can be made cross-origin without being blocked by CORS. Since forms have been able to make cross-origin requests since the early days of the web, it's important for compatibility that they should still be able to make cross-origin requests. This is why we need to implement other strategies to defend forms against CRSF, such as using a CSRF token.
```

Author: Well... a couple of things I guess. One is that we shouldn't say "blocked by CORS" - CORS is not a security mechanism, it doesn't block anything. It's an anti-security mechanism: it allows things that would otherwise be blocked. This is such a common (and dangerous) misconception that I want us to be very careful here.

But also, I don't really like:

> Simple requests, which are the sort of request that result from a `<form>` element submission, can be made cross-origin...

...because the promise of this para is defining "simple request", and in that sentence the definition is off to the side.

I could say:

> Simple requests are the sorts of requests that a `<form>` element can make. Since the early days of the web, forms have been able to make cross-origin requests, and for compatibility with legacy sites they are still allowed to. This means that sites that accept state-changing requests that could have been made by a form - that is, simple requests - must defend against CSRF using a mechanism such as a CSRF token. 

but I'm not sure this is better!

Reviewer: Fair enough. I don't like form elements making requests either - though that's a bit pedantic. My version still works if you dump mention of CORS.

```suggestion
Simple requests, which are the sort of request that result from a `<form>` element submission, can be made cross-origin without being blocked. Since forms have been able to make cross-origin requests since the early days of the web, it's important for compatibility that they should still be able to make cross-origin requests. This is why we need to implement other strategies to defend forms against CRSF, such as using a CSRF token.
```

If it were up to me I'd omit this bit about forms altogether, except perhaps as a note at the every end explaining why forms are a problem for simple requests. Something like:

```
### Avoiding simple requests

Web browsers distinguish two sorts of HTTP requests: [_simple_ requests](/en-US/docs/Web/HTTP/CORS#simple_requests) and other requests.

Simple requests, which are the sort of request that can be made cross-origin without being blocked. 
Some parts of the web platform, in particular JavaScript APIs like {{domxref("Window.fetch()", "fetch()")}}, can make different sorts of requests (for example, requests that set custom headers), and these requests are by default not allowed cross-origin, so a CSRF attack would not succeed.

So a website that uses `fetch()` or `XMLHttpRequest` can defend against CSRF by ensuring that the state-changing requests that it issues are never simple requests.

For example, setting the request's {{httpheader("Content-Type")}} to `"application/json"` will prevent it from being treated like a simple request:
```

Author: I think forms are important because they are the reason simple requests exist, and essentially are the definition for them. Without forms the story is "for some reason, browsers allow some requests cross-site". Which makes no sense, why would browsers do that?

Author: > My version still works if you dump mention of CORS.

Well, except, as I said, it still has this on-the-side definition which I don't very much like: "Simple requests, which are the sort of request that result from a `<form>` element submission, can be made cross-origin without being blocked." rather than the direct definition "Simple requests are the sort of request that result from a `<form>` element submission."

Reviewer: Yeah. It's not worth blocking all this other good stuff on. Are you going to try a new PR or should I add a note to my list to look at this bit at some point?

Thanks for all this. So much better.

Author: I applied your suggestion before merging, but if there's anything else you'd like to add, please file an issue (or a PR).

In general, a CSRF attack is possible if your website:

- Uses HTTP requests to change some state on the server.
- Uses only cookies to validate that the request came from an authenticated user.
- Uses only parameters in the request that an attacker can predict.

## Defenses against CSRF

In this section we'll outline three alternative defenses against CSRF and a fourth practice which can be used to provide defense in depth for either of the others.

- The first primary defense is to [use _CSRF tokens_](#csrf_tokens) embedded in the page. This is the most common method if you're issuing state-changing requests from form elements, as in our example above.

> - The second is to [use _Fetch metadata_](fetch_metadata) HTTP headers to check whether the start-changing request is being issued cross-site or not.

Reviewer: Maybe

```suggestion
- The second is to [use _Fetch metadata_](fetch_metadata) HTTP headers to check whether or not the start-changing request is being issued cross-site.
```
